enhancements/scorecard-storage.md

```diff
@@ -0,0 +1,269 @@
+---
+title: scorecard-persistent-storage-option
+authors:
+  - "@jemccorm"
+reviewers:
+  - TBD
+  - "@estroz"
+  - "@jesusr"
+  - "@bparees"
+approvers:
+  - TBD
+creation-date: 2021-03-26
+last-updated: 2021-04-12
+status: implementable
+see-also:
+  - "/enhancements/this-other-neat-thing.md"  
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# scorecard-storage-option
+
+This enhancement provides scorecard users a means to persist reports or other output from their scorecard tests to a persistent location that can be collected
+by scorecard.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+None
+
+## Summary
+
+QE and other test writers sometimes require storage within their test environment (e.g. test container) to write reports and other test output material.  Today within scorecard this is not possible since a persistent volume is not available within the test container.  As the test completes, the container ends, leaving only the container log on the k8s cluster until the Pod is deleted.
+
+This enhancement would offer an optional mechanism whereby test writers have a persistent volume mounted within their test containers, provisioned, enabled, and cleaned up by scorecard.
+
+
+## Motivation
+
+Allowing scorecard tests a place to persist test output is key to running advanced tests and supporting complex test output.
+
+QE teams for example might already have tests which produce test output files, in order for them to integrate their tests into scorecard, they need a persistent volume to write test results into.
+
+This feature enables a wider adoption of tests that can be executed within scorecard.
+
+### Goals
+
+Goals of this proposal include:
+* provide a persistent volume for scorecard test developers
+* Collect persistent volume output when scorecard completes
+
+### Non-Goals
+
+This enhancement doesn't change scorecard's test result format (v1alpha3) so this is not a breaking change.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+A scorecard user can specify for a given test that they want persistent storage to be made available to that test.  This is specified within the scorecard configuration file as a PersistentVolume spec.  For example, a user would specify they want storage for a test by specifying
+in the scorecard configuration file the `storage` spec as in the following example:
+
+```yaml
+  tests:
+  - image: quay.io/custom-scorecard-example/example-test:latest
+    entrypoint:
+    - example-scorecard-test
+    - example-test
+    labels:
+      suite: example
+      test: example-test
+    storage: 
```

Comment:
It would be convenient to allow this block to be specified globally in addition to per-test, to avoid repetition. The per-test block would override the global one.
```yaml
storage: 
  spec:
    storageClassName: hostpath
    capacity:
      storage: 1Gi
    accessModes:
    - ReadWriteOnce
tests:
- image: quay.io/custom-scorecard-example/example-test:latest
...
```

Reply:
I like that idea, that being to allow a global storage spec to be defined and used for ALL tests that are selected.  If the storage spec if found on the per-test level, then it would override the global setting.  I didn't do that with this feature request because I assumed that most test writers will not need or require persistent storage, but I do see the value of the feature.

Reply:
I added the ability to specify global storage just now as a distinct user story to call that feature out.